lib/smart_app_launch/endpoints/mock_smart_server.rb

```diff
@@ -213,5 +223,137 @@ def update_response_for_invalid_assertion(response, error_message)
+      when CONFIDENTIAL_ASYMMETRIC_TAG
+        key_set_input = Inferno::Repositories::SessionData.new.load( 
+          test_session_id: result.test_session_id, name: 'smart_jwk_set'
+        )
+        return confidential_asymmetric_authenticated?(request, response, key_set_input)
+      when CONFIDENTIAL_SYMMETRIC_TAG
+        client_secret_input = Inferno::Repositories::SessionData.new.load( 
+          test_session_id: result.test_session_id, name: 'smart_client_secret'
+        )
+        return confidential_symmetric_authenticated?(request, response, client_id, client_secret_input)
+      when PUBLIC_TAG
+        return true
+      end
+    end
+
+    def confidential_asymmetric_authenticated?(request, response, jwks)
+      assertion = request.params[:client_assertion]
+      if assertion.blank?
+        update_response_for_invalid_assertion(
+          response, 
+          'client_assertion missing from confidential asymmetric client request'
+        )
+        return false
+      end
+
+      signature_error = smart_assertion_signature_verification(assertion, jwks)
+
+      if signature_error.present?
+        update_response_for_invalid_assertion(response, signature_error)
+        return  false
+      end
+      
+      true
+    end
+
+    def confidential_symmetric_authenticated?(request, response, client_id, client_secret)
+      auth_header_value = request.headers['authorization']
+      error = confidential_symmetric_header_value_error(auth_header_value, client_id, client_secret)
+      if error.present?
+        update_response_for_invalid_assertion(response, error)
+        return false
+      end
+      
+      true
+    end
+
+    def confidential_symmetric_header_value_error(authorization_header_value, client_id, client_secret)
+      unless authorization_header_value.present?
+        return 'authorization header missing from confidential symmetric client request'
+      end
+      unless authorization_header_value.start_with?('Basic ')
+        return 'authorization header for confidential symmetric client request does not use Basic auth'
+      end
+      
+      client_and_secret = 
+        begin
+          Base64.strict_decode64(authorization_header_value.delete_prefix('Basic '))
+        rescue
+          return 'Basic authorization header could not be decoded'
+        end
+      expected_client_and_secret = "#{client_id}:#{client_secret}"
+      unless client_and_secret == expected_client_and_secret
+        return 'basic authorization header has the wrong decoded value - ' \
+               "expected '#{expected_client_and_secret}', got '#{client_and_secret}'"
+      end
+
+      nil
+    end
+
+    def pkce_error(verifier, challenge, method)
+      if verifier.blank?
+        'pkce check failed: no verifier provided'
+      elsif challenge.blank?
+        'pkce check failed: no challenge code provided'
+      elsif method == 'plain'
+        return nil unless challenge != verifier
```

Comment:
`plain` PKCE support is not allowed:

> SMART servers SHALL support the S256 code_challenge_method and SHALL NOT support the plain method.

 https://hl7.org/fhir/smart-app-launch/app-launch.html#considerations-for-pkce-support

Reply:
removed in [742af51](https://github.com/inferno-framework/smart-app-launch-test-kit/pull/96/commits/742af51aa15e06b30cc094a010658d3b287e036e)